[NFC] Decompose function input types

CodaFi opened the pull request:
Prepares the AST for future work to eliminate `getInput()` and perform function type argument matching in a less ad-hoc manner.

CodaFi commented:
@swift-ci please test

CodaFi commented:
@swift-ci please test macOS platform

CodaFi commented:
@swift-ci please test macOS platform

CodaFi commented:
@swift-ci please test macOS platform

DougGregor commented on include/swift/AST/Types.h:
```diff
@@ -2282,7 +2282,35 @@ getSILFunctionLanguage(SILFunctionTypeRepresentation rep) {
+  Identifier Label;
+  
+  /// Whether the parameter has a default argument.  Not valid for arguments.
+  bool HasDefaultArgument = false;
```
I commented on this on your other PR that builds on this... `HasDefaultArgument` really shouldn't be in the canonicalization. I don't think we can hoist `CallArgParam` up to the type system like this.

CodaFi replied:
We may need it for argument matching on overloads.

DougGregor replied:
It doesn't belong in the type, though, and we don't want it to look like it has meaning there. I think `CallArgParam` needs to stay where it is, and we can introduce a `AnyFunctionType::Param` (or similar) that only captures what's actually in the type system.

DougGregor commented on include/swift/AST/Types.h:
```diff
@@ -2295,7 +2323,8 @@ getSILFunctionLanguage(SILFunctionTypeRepresentation rep) {
 /// represented at the binary level as a single function pointer.
 class AnyFunctionType : public TypeBase {
-  const Type Input;
+  const ArrayRef<CallArgParam> Input;
```
As noted before, we should use tail allocation to store the parameters and save ourselves the pointer/indirection.

CodaFi replied:
`llvm::TrailingObjects` requires this class be `final`.  I can manually tail-allocate if you still want it.

DougGregor replied:
Yeah, you'd need to do the `TrailingObjects` thing on the subclasses

CodaFi committed:
Decompose function input types

Prepares the AST for future work to eliminate `getInput()`
and perform function type argument matching in a less
ad-hoc manner.

CodaFi commented:
@swift-ci please test

CodaFi commented:
@swift-ci please test

DougGregor commented on include/swift/AST/Types.h:
```diff
@@ -2501,22 +2534,35 @@ END_CAN_TYPE_WRAPPER(AnyFunctionType, Type)
 /// For example:
 ///   let x : (Float, Int) -> Int
-class FunctionType : public AnyFunctionType {
+class alignas(AnyFunctionType::Param) FunctionType final :
```
But... fixing the alignment of `FunctionType` doesn't help the trailing objects be properly aligned. You don't need this `alignas` change.

CodaFi replied:
Same for `GenericFunctionType`?

DougGregor replied:
Right.

DougGregor commented on include/swift/AST/Types.h:
```diff
@@ -2501,22 +2534,35 @@ END_CAN_TYPE_WRAPPER(AnyFunctionType, Type)
+    private llvm::TrailingObjects<FunctionType, AnyFunctionType::Param> {
+  friend TrailingObjects;
+      
+  unsigned NumParams;
```
I guess this doesn't fit into `AnyFunctionTypeBitfields`, eh? ;)

CodaFi replied:
Encode it in a pointer union and you've got a deal.

DougGregor commented on include/swift/AST/Types.h:
```diff
@@ -2587,25 +2633,35 @@ std::string getParamListAsString(ArrayRef<CallArgParam> parameters);
+    private llvm::TrailingObjects<GenericFunctionType, AnyFunctionType::Param> {
+  friend TrailingObjects;
+      
+  unsigned NumParams;
```
Sink `NumParams` down into `AnyFunctionType`?

CodaFi replied:
👍

DougGregor reviewed (approved):
I have minor comments, but this is looking great.

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
⛵️
